pallets/oracle/src/lib.rs

```diff
@@ -310,28 +322,91 @@ impl<T: Config> Pallet<T> {
 		amount: BalanceOf<T>,
 		currency_id: CurrencyId,
 	) -> Result<BalanceOf<T>, DispatchError> {
-		let rate = Self::get_price(OracleKey::ExchangeRate(currency_id))?;
-		let converted = rate.checked_mul_int(amount).ok_or(ArithmeticError::Overflow)?;
-		Ok(converted)
+		// Rate from asset to USD
+		let asset_rate = Self::get_price(OracleKey::ExchangeRate(currency_id))?;
+		// Rate from USD to USD is 1
+		let usd_rate = UnsignedFixedPoint::<T>::one();
+
+		Self::convert_amount(
+			amount,
+			asset_rate,
+			usd_rate,
+			T::DecimalsLookup::decimals(currency_id),
+			USD_DECIMALS,
+		)
 	}
 
 	pub fn usd_to_currency(
 		amount: BalanceOf<T>,
 		currency_id: CurrencyId,
 	) -> Result<BalanceOf<T>, DispatchError> {
-		let rate = Self::get_price(OracleKey::ExchangeRate(currency_id))?;
-		if amount.is_zero() {
+		// Rate from asset to USD
+		let asset_rate = Self::get_price(OracleKey::ExchangeRate(currency_id))?;
+		// Rate from USD to USD is 1
+		let usd_rate = UnsignedFixedPoint::<T>::one();
+
+		Self::convert_amount(
+			amount,
+			usd_rate,
+			asset_rate,
+			USD_DECIMALS,
+			T::DecimalsLookup::decimals(currency_id),
+		)
+	}
+
+	fn convert_amount(
+		from_amount: BalanceOf<T>,
+		from_price: T::UnsignedFixedPoint,
+		to_price: T::UnsignedFixedPoint,
+		from_decimals: u32,
+		to_decimals: u32,
+	) -> Result<BalanceOf<T>, DispatchError> {
+		if from_amount.is_zero() {
 			return Ok(Zero::zero())
 		}
 
-		// The code below performs `amount/rate`, plus necessary type conversions
-		Ok(T::UnsignedFixedPoint::checked_from_integer(amount)
-			.ok_or(Error::<T>::TryIntoIntError)?
-			.checked_div(&rate)
-			.ok_or(ArithmeticError::Underflow)?
-			.truncate_to_inner()
-			.ok_or(Error::<T>::TryIntoIntError)?
-			.unique_saturated_into())
+		let from_amount = T::UnsignedFixedPoint::checked_from_integer(from_amount)
```

Comment:
Could we run into overflows here?

This will multiply `from_amount` with 10^18. Depending on the amount of decimals of the currency (I think it should not be more than 18 decimals), we could have therefore already 36 decimals at this point for the `from_amount`. Let's say, we want to safely support amounts of at least 1 million (we should of course support more), then this would be the number 10 ** 42.

This will then be bigger than the maximum value of `u128`, which is about 10 ** 39.

I think we should get the same numerical accuracy here without the danger of overflows by just using `from_inner` on this line and `into_inner` on line 385 and 404.

Reply:
This is a very good point. I would have tried to avoid using directly `into_inner` as an intermediate representation because it may be confusing to have everything divided by `10**18`, but it is the only way to achieve representing as large values as `Balance` can. 

I wonder, is it any benefit to use `UnsignedFixedPoint` here instead of using the inner (u128) directly? I understand this may be in part because the prices are given by the oracle as `UnsignedFixedPoint` already, but perhaps an alternative is to convert the prices to `Balance` type first.

Reply:
The benefit of `UnsignedFixedPoint` is that it uses a safe multiplication function because it first converts all numbers to some 256 bit number, then multiplies them, then divides by 10 ** 18, then truncates to 128 bit. This way no overflow for the intermediate value will occur.

We would probably want to use that when multiplying with the price as that one also uses 18 decimals (we could of course leave it at 12 decimals as reported by DIA if we don't need to widen it to `UnsignedFixedPoint`, but it is probably still safer this way).

Reply:
Okay, I didn't know it converted the intermediate number during multiplication to a larger space. Then it makes sense!

Reply:
If I am not mistaken: [`checked_mul`](https://github.com/paritytech/substrate/blob/98f2e3451c9143278ec53c6718940aeabcd3b68a/primitives/arithmetic/src/fixed_point.rs#L855-L868) calls [`multiply_by_rational_with_rounding`](https://github.com/paritytech/substrate/blob/98f2e3451c9143278ec53c6718940aeabcd3b68a/primitives/arithmetic/src/helpers_128bit.rs#L187-L215), which calls [`product_of` of `Double128`](https://github.com/paritytech/substrate/blob/98f2e3451c9143278ec53c6718940aeabcd3b68a/primitives/arithmetic/src/helpers_128bit.rs#L125-L147), which returns a 256 bit value.

Reply:
Thanks for pointing it out @TorstenStueber. You are right that this change makes us support handling amounts up to 10**18 times larger. I [created a new test case](https://github.com/pendulum-chain/spacewalk/pull/490/commits/e427fbd43f55766dfecd95fbcbb44225b161a2bc) to confirm that the conversion works in the expected range of amounts.